### PR TITLE
Ogury: port new adapter from Go

### PR DESCRIPTION
**Pull request by zxPhoenix**
### 🔧 Type of changes
- [x] new bid adapter
- [ ] bid adapter update
- [ ] new feature
- [ ] new analytics adapter
- [ ] new module
- [ ] module update
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [ ] dependency update
- [ ] tech debt (test coverage, refactorings, etc.)

### ✨ What's the context?
New Ogury adapter (Port PR from PBS-Go)
[#3700](https://github.com/prebid/prebid-server-java/issues/3700)

### 🧠 Rationale behind the change
Why did you choose to make these changes? Were there any trade-offs you had to consider?

### 🔎 New Bid Adapter Checklist
- [ ] verify email contact works (I believe this happened go-side. this is a port)
- [x] NO fully dynamic hostnames
- [x] geographic host parameters are NOT required
- [x] direct use of HTTP is prohibited - *implement an existing Bidder interface that will do all the job*
- [ ] if the ORTB is just forwarded to the endpoint, use the generic adapter - *define the new adapter as the alias of the generic adapter*
- [x] cover an adapter configuration with an integration test

### 🧪 Test plan
How do you know the changes are safe to ship to production?
Unit tests with decent coverage

### 🏎 Quality check
- [x] Are your changes following [our code style guidelines](https://github.com/prebid/prebid-server-java/blob/master/docs/developers/code-style.md)?
- [ ] Are there any breaking changes in your code?
- [x] Does your test coverage exceed 90%?
- [ ] Are there any erroneous console logs, debuggers or leftover code in your changes?

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+        final List<BidderError> errors = new ArrayList<>();
+
+        if (request == null || CollectionUtils.isEmpty(request.getImp())) {
+            errors.add(BidderError.badInput("There are no valid impressions to create bid request to ogury bidder"));
+            return Result.withErrors(errors);
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:64]: No need for this checks

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                    final Imp modifiedImp = modifyImp(imp, bidFloor, modifiedImpExt);
+
+                    if (hasOguryParams(impExtBidderHoist)) {
+                        impWithOguryParamsHandler.accept(modifiedImp);
+                    }
+
+                    return modifiedImp;
+                }))
+                .map(Stream::toList).orElse(null);
+    }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:133]: Please, see examples how to iterate over imps in other adapters
See https://github.com/prebid/prebid-server-java/pull/3768/files as a starting point

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+                            .add(HttpUtil.ACCEPT_LANGUAGE_HEADER, lang != null ? lang : "en-US");
+
+                    Optional.of(device)
+                            .map(Device::getIp)
+                            .ifPresent(ip -> headers.add(HttpUtil.X_FORWARDED_FOR_HEADER, ip));
+
+                    Optional.of(device)
+                            .map(Device::getIpv6)
+                            .ifPresent(ip -> headers.add(HttpUtil.X_FORWARDED_FOR_HEADER, ip));
+                }, () -> headers.add(HttpUtil.ACCEPT_LANGUAGE_HEADER, "en-US"));
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:273]: Please, see other examples how to do it properly. For example `KrushmediaBidder`

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/resources/static/bidder-params/ogury.json
[the remark this answers is not in the document]
Comment (zxPhoenix) [reply src/main/resources/static/bidder-params/ogury.json:1]: I can't specify required fields because they are optional. Ogury requires or `assetKey and adUnitId` or `request.Site.Publisher.ID`. That is why I check those fields manually in the code. GO implementation works the same way. You can read explanation here: https://github.com/prebid/prebid-server/pull/4082/files#r1916370148

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    public OguryBidder(String endpointUrl, CurrencyConversionService currencyConversionService, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.currencyConversionService = Objects.requireNonNull(currencyConversionService);
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:55]: Use `Collections.singletonList` instead

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            try {
+                final ObjectNode impExt = resolveImpExt(imp);
+                final ObjectNode impExtBidderHoist = resolveImpExtBidderHoist(impExt);
+
+                final Imp modifiedImp = modifyImp(imp, bidRequest, impExtBidderHoist);
+                modifiedImps.add(modifiedImp);
+
+                if (hasOguryParams(impExtBidderHoist)) {
+                    impsWithOguryParams.add(modifiedImp);
+                }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:71]:
```
                 final Imp modifiedImp = modifyImp(imp, bidRequest);

                 modifiedImps.add(modifiedImp);
                 if (hasOguryParams(imp)) {
                     impsWithOguryParams.add(modifiedImp);
                 }
```

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        final boolean isKeysValid = validateRequestKeys(bidRequest, impsWithOguryParams);
+        if (!isKeysValid) {
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:78]: Inline it. Also rename method to `isValid*`

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        return impExtBidderHoist != null
+                && impExtBidderHoist.has(PREBID_FIELD_ASSET_KEY)
+                && impExtBidderHoist.has(PREBID_FIELD_ADUNIT_ID);
+    }
+
+    private boolean validateRequestKeys(BidRequest request, List<Imp> impsWithOguryParams) {
+        return !CollectionUtils.isEmpty(impsWithOguryParams) && Optional.ofNullable(request.getSite())
+                .map(Site::getPublisher)
+                .map(Publisher::getId)
+                .isEmpty();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:154]: `isPresent`

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        final MultiMap headers = HttpUtil.headers().add(HttpUtil.X_OPENRTB_VERSION_HEADER, "2.5");
+
+        if (device != null) {
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER, device.getUa());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.ACCEPT_LANGUAGE_HEADER, device.getLanguage());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIp());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIpv6());
+        } else {
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.ACCEPT_LANGUAGE_HEADER, "en-US");
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:167]: why are you trying to add this header if there is no device? I don't see such logic in GO

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+    private boolean validateRequestKeys(BidRequest request, List<Imp> impsWithOguryParams) {
+        return !CollectionUtils.isEmpty(impsWithOguryParams) && Optional.ofNullable(request.getSite())
+                .map(Site::getPublisher)
+                .map(Publisher::getId)
+                .isEmpty();
+    }
+
+    private MultiMap resolveHeaders(Device device) {
+        final MultiMap headers = HttpUtil.headers().add(HttpUtil.X_OPENRTB_VERSION_HEADER, "2.5");
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:158]: I don't see such header in GO

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        return headers;
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final String body = httpCall.getResponse().getBody();
+            if (StringUtils.isEmpty(body)) {
+                return Result.empty();
+            }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:178]: No need for this check

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        }
+
+        return switch (markupType) {
+            case 1 -> BidType.banner;
+            case 2 -> BidType.video;
+            case 3 -> BidType.audio;
+            case 4 -> BidType.xNative;
+            default -> throw new PreBidException(
+                    "Unsupported MType '%d', for impression '%s'".formatted(bid.getMtype(), bid.getImpid()));
+        };
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:228]:
```
        return switch (markupType) {
             case 1 -> BidType.banner;
             case 2 -> BidType.video;
             case 3 -> BidType.audio;
             case 4 -> BidType.xNative;
             case null, default -> throw new PreBidException(
                     "Unsupported MType '%d', for impression '%s'".formatted(markupType, bid.getImpid()));
         };
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java (hunk 10 above)
Comment (zxPhoenix) [reply src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java:178]: But, if Ogury responses with non 200 status code and with empty body then we will produce and catch exception and report server error instead of just return empty result

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+    private OguryBidder target;
+
+    @BeforeEach
+    public void setUp() {
+        target = new OguryBidder(ENDPOINT_URL, currencyConversionService, jacksonMapper);
+    }
+
+    @Test
+    public void shouldFailOnInvalidEndpointUrl() {
+        Assertions.assertThatIllegalArgumentException()
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:61]: Use static import here

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+                .method(HttpMethod.POST)
+                .uri(ENDPOINT_URL)
+                .headers(expectedHeaders)
+                .impIds(Set.of("imp_id"))
+                .body(jacksonMapper.encodeToBytes(modifiedBidRequest))
+                .payload(modifiedBidRequest)
+                .build());
+        assertThat(result.getValue()).usingRecursiveComparison().isEqualTo(expectedResult.getValue());
+        assertThat(result.getErrors()).isEmpty();
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:91]: No need for this test. Create 4 separate tests for each header and 1 test that checks that ipv6 will override ipv4 for X_FORWARDED_FOR_HEADER

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+                bidRequest.imp(List.of(Imp.builder().ext(bidder).build())));
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidrequest);
+
+        // then
+        assertThat(result.getErrors()).isNotEmpty()
+                .contains(BidderError.badInput(
+                        "Invalid request. assetKey/adUnitId or request.site.publisher.id required"));
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:108]: Add more tests that cover scenarios:
- impWithOguryParams + site.publisher.id + impWithoutOguryParams
- invalid impWithOguryParams  + site.publisher.id + impWithoutOguryParams
- impWithoutOguryParams
- imp.tagId copied from imp.id
- price tests (for each edge case)
- test that checks imp.ext fields aren't lost, and new fields are added/replaced by impExtBidderHoist fields

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+        // then
+        assertThat(result.getErrors()).hasSize(1)
+                .allSatisfy(error -> {
+                    assertThat(error.getType()).isEqualTo(BidderError.Type.bad_server_response);
+                    assertThat(error.getMessage())
+                            .startsWith("Failed to decode: Unrecognized token 'invalid'");
+                });
+        assertThat(result.getBids()).isEmpty();
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:162]: Remove this tests

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+        final CompositeBidderResponse result = target.makeBidderResponse(httpCall, null);
+
+        // then
+        assertThat(result.getErrors())
+                .containsExactly(BidderError.badServerResponse("Unsupported MType '10', for impression '123'"));
+        assertThat(result.getBids()).isEmpty();
+    }
+
+    @Test
+    public void makeBidderResponseShouldReturnBannerBid() throws JsonProcessingException {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:235]: test cases for audio, video, native

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java (hunk 13 above)
Comment (zxPhoenix) [reply src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:91]: Ogury should pass the both IPv4 and IPv6 headers. Go implementation does the same.

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 17]
```diff
@@ -0,0 +1,681 @@
...
+                                        .ext(ext)
+                                        .build())));
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidrequest);
+
+        // then
+        assertThat(result.getValue()).isEmpty();
+
+        assertThat(result.getErrors()).isNotEmpty()
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:231]: No need for `isNotEmpty`

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 18]
```diff
@@ -0,0 +1,681 @@
...
+                .contains(BidderError.badInput(
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:232]: Try to use `containsExactly` or `containsExactlyInAnyOrder` where possible

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 19]
```diff
@@ -0,0 +1,681 @@
...
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidrequest);
+
+        // then
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .flatExtracting(Imp::getTagid)
+                .hasSize(1)
+                .first()
+                .isEqualTo("id1");
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:260]: `containsExactly("id1")`

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 20]
```diff
@@ -0,0 +1,681 @@
...
+                                        .id("id1")
+                                        .ext(ext)
+                                        .build())));
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidrequest);
+
+        // then
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:255]: Use `.getPayload` where possible. It's just cleaner, there is no difference

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 21]
```diff
@@ -0,0 +1,681 @@
...
+                .isEqualTo("id1");
+
+        assertThat(result.getErrors()).isEmpty();
+    }
+
+    @Test
+    public void makeHttpRequestsShouldCleanImpExtWithoutLostExtraFields() {
+        // given
+        final ObjectNode extWithOguryKeys = givenExtWithBidderWithOguryKeys();
+        extWithOguryKeys.putIfAbsent("extra_field", TextNode.valueOf("extra_value"));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:269]: `extWithOguryKeys.put("extra_field", "extra_value");`

File: src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java [hunk 22]
```diff
@@ -0,0 +1,681 @@
...
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getExt)
+                .hasSize(1)
+                .first()
+                .satisfies(ext -> {
+                    assertThat(ext.get("extra_field").asText()).isEqualTo("extra_value");
+                    assertThat(ext.get("prebid")).isNull();
+                });
+
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ogury/OguryBidderTest.java:292]:
```
                .extracting(Imp::getExt)
                .hasSize(1)
                .first()
                .isEqualTo(mapper.valueToTree(Map.of(
                                  "extra_field", "extra_value",
                                  <other field-value pairs>)));
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

**Review by CTMBNara — APPROVED** [review]
